runners/local-artifact-service-java/src/main/java/org/apache/beam/artifact/local/LocalFileSystemArtifactStagerService.java

```diff
@@ -0,0 +1,226 @@
+      private File destination;
+      private FileChannel target;
+
+      @Override
+      public void onNext(PutArtifactRequest value) {
+        switch (value.getContentCase()) {
+          case NAME:
+            if (destination != null) {
+              StatusRuntimeException failure =
+                  Status.INVALID_ARGUMENT
+                      .withDescription(
+                          String.format(
+                              "Expected NAME only once as the first %s",
+                              PutArtifactRequest.class.getSimpleName()))
+                      .asRuntimeException();
+              responseObserver.onError(failure);
+              throw failure;
+            }
+            createFile(value.getName());
+            return;
+          case DATA:
+            if (destination == null) {
+              StatusRuntimeException failure =
+                  Status.INVALID_ARGUMENT
+                      .withDescription("Expected DATA only after the NAME has been provided")
+                      .asRuntimeException();
+              responseObserver.onError(failure);
+              throw failure;
+            }
+            writeData(value.getData().getData().asReadOnlyByteBuffer());
+            return;
+          case CONTENT_NOT_SET:
+          default:
+            StatusRuntimeException failure =
+                Status.INVALID_ARGUMENT
+                    .withDescription("One of NAME or DATA is required")
+                    .asRuntimeException();
+            responseObserver.onError(failure);
+            throw failure;
+        }
+      }
+
+      private void createFile(String artifactName) {
+        destination = getArtifactFile(artifactName);
+        try {
+          if (!destination.createNewFile()) {
+            String msg = String.format("Artifact with name %s already exists", artifactName);
+            StatusRuntimeException failure =
+                Status.ALREADY_EXISTS.withDescription(msg).asRuntimeException();
+            responseObserver.onError(failure);
+            throw failure;
+          }
+          target = new FileOutputStream(destination).getChannel();
+        } catch (IOException e) {
+          StatusRuntimeException status = Status.INTERNAL.withCause(e).asRuntimeException();
+          responseObserver.onError(status);
+          LOG.error(
+              "Failed to withRootDirectory file {} for artifact {} at base location {}",
+              destination,
+              artifactName,
+              stagingBase,
+              e);
+          throw status;
+        }
+      }
+
+      /**
+       * Write the contents of the provided {@link ByteBuffer} to the target {@link File}.
+       *
+       * <p>Will read starting at the current position of the {@link ByteBuffer} and consume all of
+       * the available bytes.
+       */
+      private void writeData(ByteBuffer byteBuffer) {
+        try {
+          target.write(byteBuffer);
```

Comment:
write isn't guaranteed to actually write all the bytes so you could be dropping data.

Consider using ByteString#writeTo and using a FileOutputStream instead of a FileChannel